server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/aws/AwsS3ObjectStorage.java

```diff
@@ -182,14 +181,28 @@ private ContextBuilder contextBuilder() {
                     .doOnRetry(retryContext -> createBucket(bucketName, configuration)));
         }
 
-        private void put(ObjectStorageBucketName bucketName, AwsS3AuthConfiguration configuration, Blob blob, File file) throws InterruptedException {
-            PutObjectRequest request = new PutObjectRequest(bucketName.asString(),
-                blob.getMetadata().getName(),
-                file);
+        private void uploadByFile(ObjectStorageBucketName bucketName, BlobId blobId, File file) throws InterruptedException {
+            PutObjectRequest request = new PutObjectRequest(bucketName.asString(), blobId.asString(), file);
+            upload(request);
+        }
+
+        private void uploadByBlob(ObjectStorageBucketName bucketName, Blob blob) throws InterruptedException, IOException {
+            try (InputStream payload = blob.getPayload().openStream()) {
+                PutObjectRequest request = new PutObjectRequest(bucketName.asString(),
+                    blob.getMetadata().getName(),
+                    payload,
+                    new ObjectMetadata());
+
+                upload(request);
+            }
+        }
 
-            getTransferManager(configuration)
+        private void upload(PutObjectRequest request) throws InterruptedException {
+            TransferManager transferManager = getTransferManager();
+            transferManager
                 .upload(request)
                 .waitForUploadResult();
```

Comment:
Does it means we miss a way to be more asychronous?

Can't we rather handle this as a mono?

Reply:
AWS API supports asynchronous polling progress on upload but I coudn't find a way to make it work with reactor

Reply:
I tried to use reactor + S3 event listener, it's awful. BTW there is another async SDK https://aws.amazon.com/blogs/developer/aws-sdk-for-java-2-x-released/

Reply:
see last commit #3025

Reply:
Check Upload::addProgressListener... From that callback it is easy to go to a mono.

Reply:
I applied your suggestion in #3025, See the first commit, but I breaks a concurrent test gracefully, no clear assumption on that.

Reply:
I had something like :

```
return Mono.defer(() -> Mono.just(getTransferManager(configuration).upload(request)))
                .flatMap(upload -> Mono.create(sink -> upload.addProgressListener((ProgressListener) progressEvent -> {
                    if (progressEvent.getEventType() == ProgressEventType.TRANSFER_COMPLETED_EVENT) {
                        sink.success();
                    }
                    if (progressEvent.getEventType() == ProgressEventType.TRANSFER_CANCELED_EVENT
                        || progressEvent.getEventType() == ProgressEventType.TRANSFER_FAILED_EVENT) {
                        sink.error(new ExecutionException(new RuntimeException("Could not complete upload " + progressEvent)));
                    }
                })));
```

In mind, so without changing of library...

Reply:
that's pretty same what I did in the bfe9a6e